compiler/sem/semexprs.nim

```diff
@@ -1605,21 +1605,72 @@ proc buildOverloadedSubscripts(n: PNode, ident: PIdent): PNode =
   for s in n: result.add s
 
 proc semDeref(c: PContext, n: PNode): PNode =
-  checkSonsLen(n, 1, c.config)
-  n[0] = semExprWithType(c, n[0])
-  let a = getConstExpr(c.module, n[0], c.idgen, c.graph)
-  if a != nil:
-    if a.kind == nkNilLit:
-      localReport(c.config, n, reportSem rsemDisallowedNilDeref)
-    n[0] = a
-  result = n
-  var t = skipTypes(n[0].typ, {tyGenericInst, tyVar, tyLent, tyAlias, tySink, tyOwned})
-  case t.kind:
-    of tyRef, tyPtr:
-      n.typ = t.lastSon
+  ## analyse deref such as `[]foo`, when given an nkBracketExpr converts it to
+  ## an nkDerefExpr and evals it, if given an nkDerefExpr evals it, nkError is
+  ## passed through.
+  ##
+  ## it to a  behaves differently depending upon whether
+  ## the deref target is a constant expression (static) or not (dynamic).
+  ##
+  ## Constant Expression:
+  ## - nil -> nkError
+  ## - ptr|ref type -> nkDeref with derefed value as first son
+  ## - not (ptr|ref) -> nil
+  ##
+  ## Dynamic Expression:
+  ## - ptr|ref type -> nkDeref with derefed value as first son
+  ## - not (ptr|ref) -> nil
+  
+  proc semDerefEval(c: PContext, n: PNode): PNode {.inline.} =
+    result = n # we allow mutation because we're evaluating `n`
 
+    let
+      derefTarget = semExprWithType(c, n[0])
+      targetAsConstExpr = getConstExpr(c.module, derefTarget, c.idgen, c.graph)
+      isTargetConstExpr = targetAsConstExpr != nil
+      isTargetConstNil = isTargetConstExpr and
+                         targetAsConstExpr.kind == nkNilLit
+      semmedTarget =
+        if isTargetConstExpr:
+          targetAsConstExpr
+        else:
+          derefTarget
+      typ = skipTypes(semmedTarget.typ,
+                      {tyGenericInst, tyVar, tyLent, tyAlias, tySink, tyOwned})
```

Comment:
It might be a bit of a stretch goal for now, but I thought about moving all hardcoded `skipTypes` sets like these in the constants, even if they are used only once in the specific PR. In the future, it might be easier to consolidate them, and right now you, as a PR author can give a meaningful name to it, like `const tyIgnoredInDeref = {tyGenericInst, tyVar, tyLent, tyAlias, tySink, tyOwned}`

Reply:
I added a local const and I noticed a possible issue for later investigation.